koap-core/src/commonMain/kotlin/Message.kt

```diff
@@ -580,20 +580,86 @@ sealed class Message {
+                result = 31 * result + kid.contentHashCode()
+                return result
+            }
+
+            override fun toString(): String {
+                val args = listOfNotNull(
+                    if (partialIv.size > 0) "partialIv=${partialIv.toHexString()}" else null,
+                    if (kidContext != null) "kidContext=${kidContext.toHexString()}" else null,
+                    if (kid != null) "kid=${kid?.toHexString()}" else null,
+                ).joinToString(separator = ", ")
+                return "Oscore($args)"
+            }
 
-            override fun toString(): String = "Oscore(value=${value.toHexString()})"
+            fun optionValue(): ByteArray {
+                val partialIvLength = partialIv.size
+                if (partialIvLength > 5) {
+                    throw IllegalArgumentException("Partial IV length of $partialIvLength is outside allowable range of 0..5")
+                }
+                val kidContextLength = if (kidContext != null) kidContext.size else 0
+                val kidFlag = if (kid != null) 0x08 else 0
+                val kidContextFlag = if (kidContext != null) 0x10 else 0
+                val flagBits = partialIvLength or kidFlag or kidContextFlag
+                if (flagBits == 0) {
+                    // If the OSCORE flag bits are all zero (0x00), the option value SHALL be empty
+                    return byteArrayOf()
+                }
+                var value = byteArrayOf(flagBits.toByte()) + partialIv
+                if (kidContext != null) {
+                    value += byteArrayOf(kidContextLength.toByte()) + kidContext
+                }
+                if (kid != null) {
+                    value += kid
+                }
+                return value
+            }
+
+            companion object {
+                fun fromOptionValue(value: ByteArray): Oscore {
+                    if (value.size == 0) {
+                        return Oscore(byteArrayOf(), null, null)
+                    }
+                    val flagBits = value[0].toInt()
+                    val partialIvLength = flagBits and 0x07
+                    val kidFlag = (flagBits and 0x08) != 0
+                    val kidContextFlag = (flagBits and 0x10) != 0
+                    val partialIv = value.sliceArray(1 until 1 + partialIvLength)
+                    val kidContextLength = if (kidContextFlag) value[1 + partialIvLength].toInt() else 0
+                    val kidContext = if (kidContextFlag) {
+                        value.sliceArray(2 + partialIvLength until 2 + partialIvLength + kidContextLength)
+                    } else {
+                        null
+                    }
+                    val kidOffset = (if (kidContextFlag) 2 else 1) + partialIvLength + kidContextLength
+                    val kid = if (kidFlag) value.sliceArray(kidOffset until value.size) else null
+                    return Oscore(partialIv, kidContext, kid)
```

Comment:
these are some 20+17 lines of fairly dense/complicated code, with magic flag constants, etc.
maybe they should go in their own `.kt` file to isolate the complexity, and using constants for flags and masks?